### PR TITLE
Add diagnostic for ocamlmerlin_reason

### PR DESCRIPTION
**Pull request by rgrinberg**
If it's missing, the error will be reported as a diagnostic.

### CODE REVIEW

**Comment by rgrinberg** [comment]
@giltho what do you think of this approach?

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 1]
```diff
@@ -47,35 +47,57 @@ let initializeInfo : InitializeResult.t =
   InitializeResult.create ~capabilities ~serverInfo ()
 
 let send_diagnostics rpc doc =
-  let command =
-    Query_protocol.Errors { lexing = true; parsing = true; typing = true }
+  let reason_merlin_available =
+    match Document.syntax doc with
+    | Ocaml -> true
+    | Reason -> Option.is_some (Bin.which "ocamlmerlin_reason")
```
Comment (imbsky) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:53]:
```suggestion
    | Reason -> Option.is_some (Bin.which "ocamlmerlin-reason")
```

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 2]
```diff
@@ -47,35 +47,57 @@ let initializeInfo : InitializeResult.t =
...
-        in
+  let uri = Document.uri doc |> Lsp.Uri.to_string in
+  match reason_merlin_available with
+  | false ->
+    let notif =
+      let diagnostics =
         let message =
-          Loc.print_main Format.str_formatter error;
-          String.trim (Format.flush_str_formatter ())
+          "Could not detect ocamlmerlin_reason. Please install reason"
```
Comment (imbsky) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:61]:
```suggestion
          "Could not detect ocamlmerlin-reason. Please install reason"
```

**Comment by giltho** [comment]
I like the diagnostic instead of failure, it's less noisy (especially in vscode) and it has the same clarity. 

The only remark I would make is that it probably should not fail in case the extension is not recognised. In a perfect world, it would detect the syntax, but I think it should at least default to something like `OCaml` + `Impl`. How and when the language server is activated is not described by the protocol, and the choice does not have to be made by the extension of the file. For example in VSCode, you can select the language mode to be `OCaml` on a file that is unnamed and has not been saved yet.

**Comment by rgrinberg** [comment]
> The only remark I would make is that it probably should not fail in case the extension is not recognised. In a perfect world, it would detect the syntax, but I think it should at least default to something like OCaml + Impl. How and when the language server is activated is not described by the protocol, and the choice does not have to be made by the extension of the file. For example in VSCode, you can select the language mode to be OCaml on a file that is unnamed and has not been saved yet.

Good call. I've removed the dependence on extensions. Take another look.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml (hunk 1 above)
Comment (rgrinberg) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml]: Good catch. Thanks.

**Pull request merged**